include/dice/template-library/flex_array.hpp

```diff
@@ -0,0 +1,253 @@
+					  "If extent is dynamic_extent, max_extent must not be dynamic_extent");
+
+		static constexpr size_t extent = extent_;
+		static constexpr size_t max_extent = max_extent_;
+		static constexpr bool is_dynamic_extent = extent == dynamic_extent;
+
+	private:
+		using inner_type = detail_flex_array::flex_array_inner<T, extent_, max_extent_>;
+
+	public:
+		using value_type = T;
+		using reference = value_type &;
+		using const_reference = value_type const &;
+		using pointer = value_type *;
+		using const_pointer = value_type const *;
+		using size_type = size_t;
+		using iterator = value_type *;
+		using const_iterator = value_type const *;
+		using reverse_iterator = std::reverse_iterator<iterator>;
+		using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+	private:
+		inner_type inner_;
+
+	public:
+		constexpr flex_array() noexcept = default;
+		constexpr flex_array(flex_array const &) noexcept = default;
+		constexpr flex_array(flex_array &&) noexcept = default;
+		constexpr flex_array &operator=(flex_array const &) noexcept = default;
+		constexpr flex_array &operator=(flex_array &&) noexcept = default;
+		constexpr ~flex_array() noexcept = default;
+
+		/**
+		 * Initializes the flex_array using an initializer_list
+		 *
+		 * @param init initializer list
+		 * @throws std::length_error if extent == dynamic_extend and init.size() exceeds max_size()
+		 *			or extent != dynamic_extent and init.size() != extent
+		 */
+		constexpr flex_array(std::initializer_list<value_type> const init) {
+			if constexpr (is_dynamic_extent) {
+				if (init.size() > max_size()) [[unlikely]] {
+					throw std::length_error{"flex_array::flex_array: maximum size exceeded"};
+				}
+
+				inner_.size_ = init.size();
+			} else if (init.size() != extent) [[unlikely]] {
+				throw std::length_error{"flex_array::flex_array: size mismatch"};
+			}
+
+			std::ranges::copy(init, begin());
+		}
+
+		/**
+		 * Initializes the flex_array using the range [first, last)
+		 *
+		 * @param first iterator to first element
+		 * @param last sentinel of the range
+		 * @throws std::length_error if distance(first, last) exceeds max_size()
+		 */
+		template<std::input_iterator Iter, std::sentinel_for<Iter> Sent>
+		constexpr flex_array(Iter first, Sent last) {
+			size_t ix = 0;
+			while (first != last) {
+				if (ix >= max_size()) [[unlikely]] {
+					throw std::length_error{"flex_array::flex_array: maximum size exceeded"};
+				}
+
+				inner_.data_[ix++] = *first++;
+			}
+
+			if constexpr (is_dynamic_extent) {
+				inner_.size_ = ix;
+			}
+		}
+
+		/**
+		 * Converts from a flex_array of dynamic extent to a flex_array of static extent
+		 *
+		 * @throws std::length_error if other.size() != extent
+		 */
+		template<size_t other_max_extent>
+		explicit constexpr flex_array(flex_array<value_type, dynamic_extent, other_max_extent> const &other) requires (!is_dynamic_extent) {
```

Comment:
assignment operators for both directions would also be nice.

Reply:
Kinda thought the implicit conversion would do that, but I'll add them

Reply:
Update: it does work like that, `dynamic = static` works `static = dynamic` doesn't because the ctor is `explicit`. So just like its supposed to be